iceoryx_dds/include/iceoryx_dds/internal/gateway/channel.inl

```diff
@@ -0,0 +1,116 @@
+// limitations under the License.
+
+namespace iox
+{
+namespace dds
+{
+// Typedefs
+template <typename IceoryxTerminal>
+using IceoryxTerminalPool = iox::cxx::ObjectPool<IceoryxTerminal, MAX_CHANNEL_NUMBER>;
+template <typename DDSTerminal>
+using DDSTerminalPool = iox::cxx::ObjectPool<DDSTerminal, MAX_CHANNEL_NUMBER>;
+
+// Statics
+template <typename IceoryxTerminal, typename DDSTerminal>
+IceoryxTerminalPool<IceoryxTerminal> Channel<IceoryxTerminal, DDSTerminal>::s_iceoryxTerminals = IceoryxTerminalPool();
+template <typename IceoryxTerminal, typename DDSTerminal>
+DDSTerminalPool<DDSTerminal> Channel<IceoryxTerminal, DDSTerminal>::s_ddsTerminals = DDSTerminalPool();
+
+template <typename IceoryxTerminal, typename DDSTerminal>
+inline Channel<IceoryxTerminal, DDSTerminal>::Channel(const iox::capro::ServiceDescription& service,
+                                                      const IceoryxTerminalPtr iceoryxTerminal,
+                                                      const DDSTerminalPtr ddsTerminal) noexcept
+    : m_service(service),
+      m_iceoryxTerminal(iceoryxTerminal),
+      m_ddsTerminal(ddsTerminal)
+{
+}
+
+template <typename IceoryxTerminal, typename DDSTerminal>
+inline Channel<IceoryxTerminal, DDSTerminal>::~Channel() noexcept
+{
+}
+
+template <typename IceoryxTerminal, typename DDSTerminal>
+inline Channel<IceoryxTerminal, DDSTerminal>::Channel(const Channel<IceoryxTerminal, DDSTerminal>& rhs) noexcept
+{
+    m_service = rhs.m_service;
+    m_ddsTerminal = rhs.m_ddsTerminal;
+    m_iceoryxTerminal = rhs.m_iceoryxTerminal;
+}
+
+template <typename IceoryxTerminal, typename DDSTerminal>
+inline Channel<IceoryxTerminal, DDSTerminal>& Channel<IceoryxTerminal, DDSTerminal>::
+operator=(const Channel<IceoryxTerminal, DDSTerminal>& rhs) noexcept
+{
+    m_service = rhs.m_service;
+    m_ddsTerminal = rhs.m_ddsTerminal;
+    m_iceoryxTerminal = rhs.m_iceoryxTerminal;
+}
+
+template <typename IceoryxTerminal, typename DDSTerminal>
+inline Channel<IceoryxTerminal, DDSTerminal>::Channel(Channel<IceoryxTerminal, DDSTerminal>&& rhs) noexcept
+{
+    m_service = std::move(rhs.m_service);
+    m_ddsTerminal = std::move(rhs.m_ddsTerminal);
+    m_iceoryxTerminal = std::move(rhs.m_iceoryxTerminal);
+}
+
+template <typename IceoryxTerminal, typename DDSTerminal>
+inline Channel<IceoryxTerminal, DDSTerminal>& Channel<IceoryxTerminal, DDSTerminal>::
+operator=(Channel<IceoryxTerminal, DDSTerminal>&& rhs) noexcept
+{
+    m_service = std::move(rhs.m_service);
+    m_ddsTerminal = std::move(rhs.m_ddsTerminal);
+    m_iceoryxTerminal = std::move(rhs.m_iceoryxTerminal);
+}
+
+template <typename IceoryxTerminal, typename DDSTerminal>
+inline Channel<IceoryxTerminal, DDSTerminal>
+Channel<IceoryxTerminal, DDSTerminal>::create(const iox::capro::ServiceDescription& service) noexcept
+{
+    // Create objects in the pool.
+    auto rawIceoryxTerminalPtr = s_iceoryxTerminals.create(std::forward<const iox::capro::ServiceDescription>(service));
```

Comment:
what happens if the pool is empty and does not return an object?

Shouldn't this maybe return an optional<Channel<...>> to make this visible to the user?

Reply:
It now returns an `iox::cxx::expected`